python/lib/core/dmod/core/events/base_function.py

```diff
@@ -0,0 +1,642 @@
+"""
+Classes used to organize, validate, and orchestrate function calls
+"""
+from __future__ import annotations
+
+import asyncio
+import typing
+import inspect
+import logging
+
+try:
+    from typing import ParamSpec
+    from typing import Concatenate
+except ImportError:
+    from typing_extensions import ParamSpec
+    from typing_extensions import Concatenate
+
+from dataclasses import field
+from dataclasses import dataclass
+
+from pydantic import root_validator
+
+PARAMS = ParamSpec("PARAMS")
+
+
+class Event:
+    def __init__(self, event_name: str):
+        self.__event_name: str = event_name
+
+    @property
+    def event_name(self) -> str:
+        return self.__event_name
+
+
+EVENT_HANDLER = typing.Callable[
+    Concatenate[
+        Event,
+        PARAMS
+    ],
+    typing.Union[typing.Coroutine, typing.Any]
+]
+
+
+class BasicParameter(typing.TypedDict):
+    index: int
+    name: str
+    type: typing.Optional[typing.Union[type, str]]
+    is_kwargs: typing.Optional[bool]
+    is_args: typing.Optional[bool]
+    default: typing.Optional[typing.Any]
+    required: typing.Optional[bool]
+
+
+@dataclass
+class EventFunctionParameter:
+    index: int
+    name: str
+    type: typing.Optional[typing.Union[type, str]] = field(default=None)
+    is_kwargs: typing.Optional[bool] = field(default=False)
```

Comment:
Defining defaults here is a protective measure to ensure that direct instantiation does not accidentally fill in the wrong details. It being handled well in implemented functions is nice, but that doesn't mean guardrails shouldn't be present. As for the `type` field being optional and having a default of `None`, it was probably due to just making sure all bases were covered when it came down to instantiation so that it wouldn't throw a fit if a value wasn't passed. I can't remember - been too long on that particular field.

Reply:
It just seems inconsistent.  Also, I missed it earlier, but aren't you sufficiently protecting against `None` values with the root validator function?

Reply:
Just triple checked - something like:
```python
type: typing.Optional[typing.Union[type, str]]
```

just means that `type` may be a `type`, a `str`, or `None`, but it is still required when creating the object. The following, though, allows `type` to just not be provided period:

```python
type: typing.Optional[typing.Union[type, str]] = field(default=None)
```

This is expected behavior - this entire structure needs to work where the values with `field(default=...)` can be totally omitted during initialization. `EventFunctionParameter(index=2, name="input_3")` is the most basic of the base cases. That just means that the value that goes into that index as a parameter just needs to be named `"input_3"`. Everything else is a bonus.

This entire structure is incorrect as it stands, however - this is decorated as a `dataclass` not a pydantic `BaseClass`. That fix is going to be submitted for a PR on another branch.